tracing-subscriber/src/layer/mod.rs

```diff
@@ -395,40 +395,80 @@
 //!
 //! ## Runtime Configuration With Layers
 //!
-//! Conditionally enabling/disabling layers at runtime can be challenging due to the generics
-//! used Layers implementations. However, A Layer wrapped in an `Option` is also Layers, which
-//! allows some runtime configuration:
+//! In some cases, a particular [`Layer`] may be enabled or disabled based on
+//! runtime configuration. This can introduce challenges, because the type of a
+//! layered [`Subscriber`] depends on which layers are added to it: if an `if`
+//! or `match` expression adds some [`Layer`]s in one branch and other layers
+//! in another, the [`Subscriber`] values returned by those branches will have
+//! different types. For example, the following _will not_ work:
+//!
+//! ```compile_fail
+//! # fn docs() -> Result<(), Box<dyn std::error::Error + 'static>> {
+//! # use std::path::PathBuf;
+//! # struct Config {
+//! #    is_prod: bool,
+//! #    path: PathBuf,
+//! # }
+//! # let cfg = Config { is_prod: false, path: PathBuf::from("debug.log") };
+//! use std::{fs::File, sync::Arc};
+//! use tracing_subscriber::{Registry, prelude::*};
 //!
+//! let stdout_log = tracing_subscriber::fmt::layer().pretty();
+//! let subscriber = Registry::default().with(stdout_log);
+//!
+//! // The compile error will occur here because the if and else
+//! // branches have different, and therefore incompatible, types.
+//! let subscriber = if cfg.is_prod {
+//!     let file = File::create(cfg.path)?;
+//!     let layer = tracing_subscriber::fmt::layer()
+//!         .json()
+//!         .with_writer(Arc::new(file));
+//!     subscriber.with(layer)
+//! } else {
+//!     subscriber
+//! };
+//!
+//! tracing::subscriber::set_global_default(subscriber)
+//!     .expect("Unable to set global subscriber");
+//! # Ok(()) }
 //! ```
-//! # // wrap this in a function so we don't actually create `debug.log` when
-//! # // running the doctests..
+//!
+//! However, a [`Layer`] wrapped in an [`Option`] [also implements the `Layer`
+//! trait][option-impl]. This allows individual layers to be enabled or disabled at
+//! runtime while always producing a [`Subscriber`] of the same type. For
+//! example:
+//!
+//! ```
+//! # use std::path::PathBuf;
 //! # fn docs() -> Result<(), Box<dyn std::error::Error + 'static>> {
+//! # struct Config {
+//! #    is_prod: bool,
+//! #    path: PathBuf,
+//! # }
+//! # let cfg = Config { is_prod: false, path: PathBuf::from("debug.log") };
```

Comment:
nit/tioli: this could maybe be simplified to avoid having to import `PathBuf`
```suggestion
//! # fn docs() -> Result<(), Box<dyn std::error::Error + 'static>> {
//! # struct Config {
//! #    is_prod: bool,
//! #    path: &'static str,
//! # }
//! # let cfg = Config { is_prod: false, path: "debug.log" };
```